src/etc/inc/plugins.inc.d/core.inc

```diff
@@ -82,16 +82,48 @@ function core_services()
         $services[] = $pconfig;
     }
 
-    if (is_apinger_enabled()) {
-        $pconfig = array();
-        $pconfig['name'] = "apinger";
-        $pconfig['description'] = gettext("Gateway Monitoring Daemon");
-        $pconfig['php']['restart'] = array('setup_gateways_monitor');
-        $pconfig['php']['start'] = array('setup_gateways_monitor');
-        $pconfig['pidfile'] = '/var/run/apinger.pid';
-        $services[] = $pconfig;
-    }
+    if (!isset($config['system']['prefer_dpinger'])) {
+        if (is_apinger_enabled()) {
+            $pconfig = array();
+            $pconfig['name'] = "apinger";
+            $pconfig['description'] = gettext("Gateway Monitoring Daemon");
+            $pconfig['php']['restart'] = array('setup_gateways_monitor');
+            $pconfig['php']['start'] = array('setup_gateways_monitor');
+            $pconfig['pidfile'] = '/var/run/apinger.pid';
+            $services[] = $pconfig;
+        }
+    } else {
+        if (is_dpinger_enabled()) {
 
+        $pidfiles = glob("/var/run/dpinger_*.pid");
+
+        $gateways_arr_all = return_gateways_array(true);
+
+        foreach($gateways_arr_all as $name => $gateway) {
+            if (isset($gateway['monitor_disable'])) {
+                continue;
+            }
+
+            $pattern = "/^dpinger_({$name})~([^~]+)~([^~]+)\.pid$/";
+            foreach ($pidfiles as $pidfile) {
+                if (preg_match($pattern,basename($pidfile), $matches)) {
+                    $dpinger_pid = $pidfile;
```

Comment:
wouldn't it be enough to construct the pid file here instead of glob and find ?

this loop must also take into account monitoring_disabled flag and skip that particular dpinger process because it doesn't exist

Reply:
The pid filename is created in gwlib and the name is horribly complex. At first I could see no reason for the name being as complex as it is... but after thinking about it, I can see the reason behind it,.

Reply:
which is? :)

Reply:
You may have multiple targets on the same interface.

Reply:
but we only have one interface to monitor?

Reply:
I suppose because the target is the variable?  best ask Denny on that one. :)